File: Searching/PythonCodes/Leetcode349.py
class Solution(object):
    def binarySearch(self, nums, target):
        start, end = 0, len(nums) - 1
        while start <= end:
            mid = start + (end - start) // 2
            if nums[mid] == target:
                return True
            if nums[mid] > target:
                end = mid - 1
            else:
                start = mid + 1
        return False

    def intersection(self, nums1, nums2):
        nums1.sort()
        nums2.sort()
        result = []
        didWeTrackedAlready = [False] * 10001

        for num in nums1:
            if didWeTrackedAlready[num]:
                continue
            didWeTrackedAlready[num] = True
            if self.binarySearch(nums2, num):
                result.append(num)

        return result

File: Searching/JavaCodes/Leetcode349.java
class Solution {
    public static boolean binarySearch(int[] nums,int target){
        int start=0,end= nums.length-1;
        while(start<=end){
            int mid=start+(end-start)/2;
            if(nums[mid]==target){
                return true;
            }
            if(nums[mid]>target){
                end=mid-1;
            }
            else{
                start=mid+1;
            }
        }
        return false;
    }


    public static int[] intersection(int[] a, int[] b) {
        Arrays.sort(a);
        Arrays.sort(b);

        List<Integer> list=new ArrayList<>();
        boolean[] didWeTrackedAlready=new boolean[10001];


        for (int i = 0; i < a.length; i++) {
            int target=a[i];
            if(didWeTrackedAlready[target]){
                continue;
            }
            didWeTrackedAlready[target]=true;
            if(binarySearch(b,target)){
                list.add(target);
            }
        }

        int[] result=new int[list.size()];
        for (int i = 0; i < result.length; i++) {
            result[i]=list.get(i);
        }
        return result;
    }
}